diff --git a/intel_extension_for_pytorch/quantization/_utils.py b/intel_extension_for_pytorch/quantization/_utils.py
index 08a4302..82620a2 100644
--- a/intel_extension_for_pytorch/quantization/_utils.py
+++ b/intel_extension_for_pytorch/quantization/_utils.py
@@ -594,6 +594,7 @@ def load_qconf_summary_to_model(model, qconf_summary):
                         v.tensor_id_to_scale_zp[tensor_info["id"]] = (scale, zp)
                 else:
                     input_tensor_infos.append(None)
+                    input_force_dtype_infos.append(None)
             weight_tensor_infos = []
             weight_idx = 0
             for tensor_info in q_op_info["weight_tensor_infos"]:
